h2o-core/src/main/java/water/rapids/ast/prims/reducers/AstTopN.java

```diff
@@ -47,68 +46,172 @@ public String description() {
 				final String[] _columnName;   // name of column that we are grabbing top N for
 				PriorityQueue _sortQueue;
+				long[] _rowIndices;  // store original row indices of values that are grabbed
+				long[] _lValues;         // store the grabbed values
+				double[] _dValues;  // store grabbed longs
 				Frame _sortedOut;   // store the final result of sorting
 				final int _rowSize;   // number of top or bottom rows to keep
-				final boolean _increasing;  // sort with Top values first if true.
+				final int _flipSign;   // 1 for top values, -1 for bottom values
 				boolean _csLong = false;      // chunk of interest is long
 
-				private GrabTopNPQ(String[] columnName, long rowSize, boolean increasing) {
+				private GrabTopNPQ(String[] columnName, long rowSize, int flipSign) {
 						_columnName = columnName;
 						_rowSize = (int) rowSize;
-						_increasing = increasing;
+						_flipSign = flipSign;
 				}
 
 				@Override
 				public void map(Chunk cs) {
 						_sortQueue = new PriorityQueue<RowValue<E>>(); // instantiate a priority queue
 						_csLong = cs instanceof C8Chunk;
-						Long startRow = cs.start();           // absolute row offset
+						long startRow = cs.start();           // absolute row offset
 
 						for (int rowIndex = 0; rowIndex < cs._len; rowIndex++) {  // stuff our chunks into priorityQueue
 								long absRowIndex = rowIndex + startRow;
 								if (!cs.isNA(rowIndex)) { // skip NAN values
 										addOneValue(cs, rowIndex, absRowIndex, _sortQueue);
 								}
 						}
+
+						// copy the PQ into the corresponding arrays
+						if (_csLong) {
+								_lValues = new long[_sortQueue.size()];
+								copyPQ2ArryL(_sortQueue, _lValues);
+						} else {
+								_dValues = new double[_sortQueue.size()];
+								copyPQ2ArryD(_sortQueue, _dValues);
+						}
+				}
+
+				public void copyPQ2ArryL(PriorityQueue sortQueue, long[] values) {
+						//copy values on PQ into arrays in sorted order
+						int qSize = sortQueue.size();
+						_rowIndices = new long[qSize];
+
+						for (int index = qSize - 1; index >= 0; index--) {
+								RowValue tempPairs = (RowValue) sortQueue.poll();
+								_rowIndices[index] = tempPairs.getRow();
+								values[index] = (long) tempPairs.getValue();
+						}
+				}
+
+				public <T> void copyPQ2ArryD(PriorityQueue sortQueue, double[] values) {
+						//copy values on PQ into arrays in sorted order
+						int qSize = sortQueue.size();
+						_rowIndices = new long[qSize];
+
+						for (int index = qSize - 1; index >= 0; index--) {
+								RowValue tempPairs = (RowValue) sortQueue.poll();
+								_rowIndices[index] = tempPairs.getRow();
+								values[index] = (double) tempPairs.getValue();
+						}
 				}
 
 				@Override
 				public void reduce(GrabTopNPQ<E> other) {
-						this._sortQueue.addAll(other._sortQueue);
+						// do a combine here of two arrays.  Note the value always store values that are increasing
+						if (_csLong)
```

Comment:
For my task, there is only one column per chunk.  Hence, it is only going to be either long or doubles.  There will not be mixed data types within one column, right?